active/003-seq-feature.md

```diff
@@ -0,0 +1,126 @@
+
+### enforced attributes
+Besides user arbitrarily given attributes, we enforce the following attribues:
+
+* `intervals`: store intervals of coordinates
+* `strand`: strand
+* `wref`: a weak reference to `IntervalMetadata` object
+* `boundaries`: records the openness of each interval.  So a boundary of `(True, False)` would it indicate that the exact right boundary is unknown, corresponding to the examples [here](ftp://ftp.ebi.ac.uk/pub/databases/embl/doc/FT_current.html#3.4.3).
+
+
+### methods
+`__init__(**kwargs)`
+The construction would be like:
+```python
+>>> f = BoundFeature(intervals=[1, (4, 7)], strand='+', boundaries=[(True, True), (False, False)], wref=None, gene='sagA', function='toxin')
+>>> f.intervals  # get coordinates
+>>> f.wref   # get weak ref to the interval metadata
+>>> f.function   # get the feature info
+```
+Note, in the above example, the interval `1` is shorthand for `(1, 2)`.
+
+`update(**kwargs)`
+- `**kwargs`: List of attributes that will be modified for the given `BoundFeature` object.
+- Updates the corresponding entries in the `IntervalMetadata`, if intervals are modified, through the weakref
+
+```python
+f = BoundFeature(name='sagA', function='transport')
+f.update(function='toxin')
+f.update(intervals=[(1, 2)])
+```
+
+
+## `IntervalMetadata` object
+### Attributes
+* `features`: a list of `BoundFeature` objects
+* `_staled`: boolean. Whenever any intervals of any member feature is modified, this is set to True, indicating `_intervaltree` needs to be updated.
+* `_intervaltree` (implemented as property): `IntervalTree` object created from all the intervales in the `features`. When this attributes is accessed by any methods (like `query` below), it checks `_staled` to decide whether to re-create from all intervals before it returns the interval tree. This lazy update saves computation.
+
+   This is implemented as bx-python `IntervalTree` object.  The keys correspond to intervals and the values correspond to a single `BoundFeature` object. 
+
+   Some benchmarks have been made for various IntervalTree data structures - in this [benchmark](https://gist.github.com/shoyer/c939325f509d7c027949), bx-python was determined to have one of the fastest query times.
+
+### Methods
+`__init__(self, features=None)`
+- `features` : list of `BoundFeature` objects.
+- Called to initialize the `IntervalMetadata` object
+- The weakrefs will be updated here to link each `BoundFeature` object to the current `IntervalMetadata` object
+
+`reverse_complement(length)`
```

Comment:
Copying my questions from previous discussion:

What is being complemented here? You're just reversing/swapping coordinates. Perhaps rename to `IntervalMetadata.reverse()`? Especially important if interval metadata becomes associated with non-sequence objects where "complement" has no meaning.

It's also possible that a `length` parameter isn't necessary. Since these features will be bound to a `Sequence` object (or at least an object that has some sort of length), we could use a similar approach to [`PositionalMetadataMixin._positional_metadata_axis_len_`](https://github.com/biocore/scikit-bio/blob/master/skbio/metadata/_mixin.py#L160)?